Load only the crate header for `locator::crate_matches`

jyn514 opened the pull request:
Previously, we used the following info to determine whether to load the crate:
1. The METADATA_HEADER, which includes a METADATA_VERSION constant
2. The embedded rustc version
3. Various metadata in the `CrateRoot`, including the SVH

This worked ok most of the time. Unfortunately, when building locally the rustc version is always
the same because `omit-git-hash` is on by default. That meant that we depended only on 1 and 3, and
we are not very good about bumping METADATA_VERSION (it's currently at 7) so in practice we were
only depending on 3. `CrateRoot` is a very large struct and changes somewhat regularly, so this led
to a steady stream of crashes from trying to load it.

Change the logic to add an intermediate step between 2 and 3: introduce a new `CrateHeader` struct
that contains only the minimum info needed to decide whether the crate should be loaded or not. That
avoids having to load all of `CrateRoot`, which in practice means we should crash much less often.

Note that this works because the SVH should be different between any two dependencies, even if the
compiler has changed, because we use `-Zbinary-dep-depinfo` in bootstrap. See
https://github.com/rust-lang/rust/pull/111329#issuecomment-1538303474 for more details about how the
original crash happened.

Helps with https://github.com/rust-lang/rust/issues/76720.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-14`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/4909997281/jobs/8766749753) [(plain)](https://github.com/rust-lang/rust/commit/10296b10e790f8b21983bae0ba0cfc9ecc4addaa/checks/13299224085/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
##[group]Building stage0 tool linkchecker (x86_64-unknown-linux-gnu)
Building stage0 tool linkchecker (x86_64-unknown-linux-gnu)
    Finished release [optimized] target(s) in 0.13s
##[endgroup]
nomicon/unchecked-uninit.html:159: broken link - `core/mem/union.MaybeUninit.html`
nomicon/unchecked-uninit.html:201: broken link - `core/mem/union.MaybeUninit.html`
nomicon/unchecked-uninit.html:236: broken link - `core/ptr/index.html`
nomicon/unchecked-uninit.html:238: broken link - `core/ptr/fn.write.html`
nomicon/unchecked-uninit.html:277: broken link - `core/ptr/macro.addr_of_mut.html`
nomicon/print.html:3241: broken link - `core/mem/union.MaybeUninit.html`
nomicon/print.html:3283: broken link - `core/mem/union.MaybeUninit.html`
nomicon/print.html:3318: broken link - `core/ptr/index.html`
nomicon/print.html:3320: broken link - `core/ptr/fn.write.html`
nomicon/print.html:3359: broken link - `core/ptr/macro.addr_of_mut.html`
nomicon/print.html:7126: broken link - `core/panic/struct.PanicInfo.html`
nomicon/panic-handler.html:154: broken link - `core/panic/struct.PanicInfo.html`
std/arch/index.html:149: broken link - `core/arch/x86/index.html`
std/arch/index.html:150: broken link - `core/arch/x86_64/index.html`
std/arch/index.html:151: broken link - `core/arch/arm/index.html`
std/arch/index.html:152: broken link - `core/arch/aarch64/index.html`
std/arch/index.html:153: broken link - `core/arch/riscv32/index.html`
std/arch/index.html:154: broken link - `core/arch/riscv64/index.html`
std/arch/index.html:155: broken link - `core/arch/mips/index.html`
std/arch/index.html:156: broken link - `core/arch/mips64/index.html`
std/arch/index.html:157: broken link - `core/arch/powerpc/index.html`
std/arch/index.html:158: broken link - `core/arch/powerpc64/index.html`
std/arch/index.html:159: broken link - `core/arch/nvptx/index.html`
std/arch/index.html:160: broken link - `core/arch/wasm32/index.html`
edition-guide/rust-2018/path-changes.html:217: broken link - `core/index.html`
edition-guide/print.html:554: broken link - `core/index.html`
reference/attributes/codegen.html:340: broken link - `core/panic/struct.Location.html`
reference/attributes/codegen.html:343: broken link - `core/panic/struct.Location.html`
reference/attributes/codegen.html:352: broken link - `core/panic/struct.Location.html`
reference/attributes/codegen.html:353: broken link - `core/intrinsics/fn.caller_location.html`
reference/expressions/literal-expr.html:253: broken link - `core/primitive.u128.html`
reference/expressions/literal-expr.html:306: broken link - `core/primitive.f32.html`
reference/expressions/literal-expr.html:306: broken link - `core/primitive.f64.html`
reference/expressions/literal-expr.html:314: broken link - `core/primitive.f32.html`
reference/expressions/literal-expr.html:314: broken link - `core/primitive.f64.html`
reference/expressions/literal-expr.html:314: broken link - `core/primitive.f32.html`
reference/expressions/literal-expr.html:314: broken link - `core/primitive.f64.html`
reference/types/pointer.html:192: broken link - `core/ptr/macro.addr_of.html`
reference/types/pointer.html:192: broken link - `core/ptr/macro.addr_of_mut.html`
reference/names/preludes.html:173: broken link - `core/prelude/rust_2015/index.html`
reference/names/preludes.html:174: broken link - `core/prelude/rust_2018/index.html`
reference/names/preludes.html:175: broken link - `core/prelude/rust_2021/index.html`
reference/names/preludes.html:181: broken link - `core/prelude/rust_2015/index.html`
reference/names/preludes.html:181: broken link - `core/prelude/rust_2018/index.html`
reference/names/preludes.html:181: broken link - `core/prelude/v1/index.html`
reference/names/preludes.html:187: broken link - `core/index.html`
reference/names/preludes.html:217: broken link - `core/index.html`
reference/names/preludes.html:224: broken link - `core/index.html`
reference/behavior-considered-undefined.html:255: broken link - `core/ptr/struct.NonNull.html`
reference/behavior-considered-undefined.html:255: broken link - `core/num/index.html`
reference/inline-assembly.html:155: broken link - `core/arch/macro.asm.html`
reference/inline-assembly.html:155: broken link - `core/arch/macro.global_asm.html`
reference/runtime.html:159: broken link - `core/panic/struct.PanicInfo.html`
reference/print.html:2294: broken link - `core/sync/atomic/index.html`
reference/print.html:6182: broken link - `core/panic/struct.Location.html`
reference/print.html:6185: broken link - `core/panic/struct.Location.html`
reference/print.html:6194: broken link - `core/panic/struct.Location.html`
reference/print.html:6195: broken link - `core/intrinsics/fn.caller_location.html`
reference/print.html:6907: broken link - `core/primitive.u128.html`
reference/print.html:6960: broken link - `core/primitive.f32.html`
reference/print.html:6960: broken link - `core/primitive.f64.html`
reference/print.html:6968: broken link - `core/primitive.f32.html`
reference/print.html:6968: broken link - `core/primitive.f64.html`
reference/print.html:6968: broken link - `core/primitive.f32.html`
reference/print.html:6968: broken link - `core/primitive.f64.html`
reference/print.html:10286: broken link - `core/ptr/macro.addr_of.html`
reference/print.html:10286: broken link - `core/ptr/macro.addr_of_mut.html`
reference/print.html:12446: broken link - `core/prelude/rust_2015/index.html`
reference/print.html:12447: broken link - `core/prelude/rust_2018/index.html`
reference/print.html:12448: broken link - `core/prelude/rust_2021/index.html`
reference/print.html:12454: broken link - `core/prelude/rust_2015/index.html`
reference/print.html:12454: broken link - `core/prelude/rust_2018/index.html`
reference/print.html:12454: broken link - `core/prelude/v1/index.html`
reference/print.html:12460: broken link - `core/index.html`
reference/print.html:12490: broken link - `core/index.html`
reference/print.html:12497: broken link - `core/index.html`
reference/print.html:13341: broken link - `core/arch/macro.asm.html`
reference/print.html:13341: broken link - `core/arch/macro.global_asm.html`
reference/print.html:14150: broken link - `core/ptr/struct.NonNull.html`
reference/print.html:14150: broken link - `core/num/index.html`
reference/print.html:14412: broken link - `core/panic/struct.PanicInfo.html`
reference/conditional-compilation.html:321: broken link - `core/sync/atomic/index.html`
number of HTML files scanned: 4867
number of HTML redirects found: 425
number of links checked: 1520886
number of links ignored due to external: 158130
```

</details>

Mark-Simulacrum commented:
Depdep is binary dep info yeah, the flag is -Zbinary-dep-depinfo IIRC.

My sense is that this has a chance of causing extra invalidation where it's not technically needed... maybe due to e.g. doc clearing out the full build directory or something? I don't have a great test case in mind though, I'm not sure why I(?) limited that originally.

Looking at the past PRs in this area, I had cited this case:

> also avoid rebuilding the whole compiler on changes to libtest (with keep-stage to limit to a single stage build) since there isn't a dependency edge between libtest and most compiler crates. That seems like a good feature to keep.

Can you check if that breaks here? My guess is yes.

In general since I wrote that on https://github.com/rust-lang/rust/pull/86641#issuecomment-885611992, I feel like this case is less important, but would be good to have a clear sense of tradeoffs. Working builds are probably more important than slightly faster builds...

That said, my sense is that something about the depinfo changing is still a better fix here, at least in the long run. Maybe now that we have a test case that's worth taking another look at.

jyn514 commented:
> I'm not sure why I(?) limited that originally.

looks like the original PR adding this was https://github.com/rust-lang/rust/pull/67760, and you actually had the same code from this PR until you found in https://github.com/rust-lang/rust/pull/67760#issuecomment-569990164 that it broke the tests (same as in this PR :smile:). I think that test failure is fixable while still keeping this change (maybe we're deleting the sysroot too aggressively?) - I'll look into it.

> > also avoid rebuilding the whole compiler on changes to libtest (with keep-stage to limit to a single stage build) since there isn't a dependency edge between libtest and most compiler crates. That seems like a good feature to keep.
> 
> Can you check if that breaks here? My guess is yes.

It does not :)

```
; x b rustc_span --stage 0
Building stage0 library artifacts (x86_64-unknown-linux-gnu)
    Finished release [optimized] target(s) in 0.07s
Building compiler artifacts {rustc_span} (stage0 -> stage1, x86_64-unknown-linux-gnu)
   Compiling rustc_span v0.0.0 (/home/jyn/src/rust/compiler/rustc_span)
    Finished release [optimized + debuginfo] target(s) in 0.87s
; touch library/test/src/lib.rs 
; x b rustc_span --stage 0
Building stage0 library artifacts (x86_64-unknown-linux-gnu)
   Compiling test v0.0.0 (/home/jyn/src/rust/library/test)
   Compiling sysroot v0.0.0 (/home/jyn/src/rust/library/sysroot)
    Finished release [optimized] target(s) in 0.30s
Building compiler artifacts {rustc_span} (stage0 -> stage1, x86_64-unknown-linux-gnu)
    Finished release [optimized + debuginfo] target(s) in 0.09s
```

> That said, my sense is that something about the depinfo changing is still a better fix here, at least in the long run. Maybe now that we have a test case that's worth taking another look at.

Hmm, ok - I thought bindep-depinfo was only for rustbuild, but looking at https://github.com/rust-lang/rust/issues/63012#issuecomment-1214209860 I see @jonhoo and others are using it for toolchain tracking too. I'll see if I can minimize the bug.

jyn514 commented:
Actually, it looks like binary-dep-depinfo doesn't even try to address this use case? Am I missing something? https://github.com/rust-lang/rust/blob/1fa1f5932b487a2ac4105deca26493bb8013a9a6/compiler/rustc_interface/src/passes.rs#L491-L510

bjorn3 commented:
`-Zbinary-dep-depinfo` does include the standard library parts used as dependencies by the respective crate, just not rustc itself.

jyn514 commented:
I tried adding the toolchain to the dep-info file (i.e. `build/x86_64-unknown-linux-gnu/stage1/bin/rustc:` gets added to the end of every file) and it did *not* solve this problem. I think the issue is that -Zbinary-depinfo only tells cargo to *rerun* rustc, not that it needs to also clear the incremental and metadata caches.

Mark-Simulacrum commented:
I think that test is actually misleading? You're not rebuilding the compiler in the first step, so we're not running the code here at all. You need to build through to the bin/rustc binary (I forget what Step does that). After that I still expect it to cause more re-compilations than needed.

> I thought bindep-depinfo was only for rustbuild [...]
> Actually, it looks like binary-dep-depinfo doesn't even try to address this use case? Am I missing something? 

I'm not sure I understand what *this use case* is. The expectation in my mind is that *if* you change rustc in a way that changes metadata format, you should get a new std. The binary depdep info will include the dependency on std, and so we will cause a rebuild of the compiler.

I think that isn't true with keep-stage1-std or whatever that flag is, but I'm not sure that the fully right fix is trying to fix that... I guess we can make this conditional on that flag? Essentially if we've artificially kept any dependency Step the next one needs this special cleaning logic. (For example, keeping rustc would mean needing this logic for rustdoc, if I'm connecting the dots right?)

Essentially keep-stage bypasses this logic by avoiding an update to the Std/Library binaries.

But with https://github.com/rust-lang/rust/issues/76720#issuecomment-1537546470 I guess we're seeing this without keep-stage? I'm not sure what is particular about that test case...

> it needs to also clear the incremental and metadata caches.

I'm not sure about incremental, but why would metadata caches be a problem? rustc should never be loading metadata it's not compatible with because dependencies are built by Cargo *before* the next rustc runs.

---

Looking at the test case in https://github.com/rust-lang/rust/issues/76720#issuecomment-1537546470 we're failing to recompile libserde_derive in stage1-tools for some reason, that's the file we fail to load:

```
INFO rustc_metadata::locator dylib reading metadata from: /home/mark/Build/rust-3/build/x86_64-unknown-linux-gnu/stage1-tools/release/deps/libserde_derive-5970f6e0b3790395.so
DEBUG rustc_metadata::locator checking 8 bytes of metadata-version stamp
DEBUG rustc_metadata::locator inflating 25684 bytes of compressed metadata
```

I'm going to try to figure out why we're not recompiling libserde_derive -- the .d file rustc emits for it points to a non-existent libstd file: `/home/mark/Build/rust-3/build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd-c38e6f867e4a21ae.rlib`, so I would expect Cargo to do the right thing here without further changes to rustbuild, but that doesn't seem to be happening.

Mark-Simulacrum commented:
https://gist.githubusercontent.com/Mark-Simulacrum/3e5d4dc8d659da79376d1181617d121a/raw/3b5176553de7ec6d0b7649bc1e0552c84dfee0f2/log is the log file including Cargo's fingerprint and rustc's metadata loading. I'm still trying to dig through that, no clear leads just yet.

My latest suspicion is that Cargo actually *does* rebuild the serde_derive and serde crates (in an earlier invocation), but that when *rustc* goes to look for the libserde_derive crate it's failing to load that as it's loading an earlier version of that crate. (Cargo doesn't delete stale files in the target directory).

I'm not quite sure yet why we see this only with serde-derive and not other (non proc macro?) dependencies...

Mark-Simulacrum commented:
OK, so I think I have at least a theoretical understanding of what causes this. You need to go from version A to version B of rust-lang/rust where both of these are true (plus having built A):

* Some dependency version changed (e.g., serde_derive 1.0.x to 1.0.y).
* The compiler's metadata format changed, without bumping the metadata version constant.
   * And you're using ignore-git or similar, so the rustc version information isn't enough to reject the crate either, causing us to go load the full CrateRoot.

The first condition is needed so that when rustc is run, it outputs a distinct/separate file (e.g., in this case we have version B with libserde_derive-67ce481370f50db2.so and version A with libserde_derive-5970f6e0b3790395.so). The second condition is needed so that when we go to load the metadata from each of these to get the full SVH and confirm which one is correct, we fail to deserialize the metadata at that time (this is what causes the ICE).

I *think* a better fix than the one in this PR is to change our metadata format so that we can get at the SVH, crate name, etc. (i.e., everything accessed here: https://github.com/rust-lang/rust/blob/0dddad0dc5d614f799d7e04de4895e7a7418eccb/compiler/rustc_metadata/src/locator.rs#L668-L701) without needing to deserialize a bunch of the other information in the root (even just at Lazy level). This should be pretty easy to do, and would mean that the metadata constant only needs to be bumped when a small set of information changes, rather than abstractly when "something" changes, which we're bad at getting right.

(I think it should even be possible to make the metadata version something we compute dynamically, e.g., by serializing a known set of information in this "CratePrefix" and then hashing that, which may be an alternative fix here).

Either of these approaches would mean that the ICE hit in https://github.com/rust-lang/rust/issues/76720#issuecomment-1537546470 isn't hit anymore, while preserving the cache behavior we want (around libtest -> rustc_span not being a valid edge, for example). This should also make it much more rare that this happens outside of rustc tree, which I think is sometimes the case today, particularly where others also rely on -Zbinary-dep-depinfo, which we're hoping to stabilize "soon".

@petrochenkov -- do you have thoughts on the metadata format change here? Preference for whether to change the way we compute the version constant or moving some information into a crateprefix struct or similar with more "stable" encoding (e.g., no optimization around laziness, etc.)?

petrochenkov commented:
@Mark-Simulacrum 
>do you have thoughts on the metadata format change here?

I barely remember what is this all about.
I also never changed the version constant and don't know when it's supposed to be changed.
@bjorn3 also tried to make some changes that may be related to this issue in https://github.com/rust-lang/rust/pull/93945.

bjorn3 commented:
I think changing the crate metadata format to include all information used by the `CrateLocator::crate_matches` function (with the exception of if it is a proc macro or not I think) in the metadata header next to the rustc version makes sense. It could be useful if I revive https://github.com/rust-lang/rust/pull/93945 too.

That being said, even with that change we will still need to manually clear the incremental cache, but at least that shouldn't cause any unnecessary recompilations. Just slower recompilations when a recompilation is actually necessary.

jyn514 commented:
I think an easier solution for incremental is to just turn it off altogether for stage 1+ builds; it seems unlikely in practice that we'll ever actually be able to reuse the cache.

bjorn3 commented:
With `--keep-stage` enabling incremental for stage > 0 should help. But `--keep-stage` can cause this kind of compiler crashes anyway, so maybe not worth optimizing for that.

jyn514 commented:
> OK, so I think I have at least a theoretical understanding of what causes this. You need to go from version A to version B of rust-lang/rust where both of these are true (plus having built A):

this was exactly it!!

here's a script to reproduce it (assuming you have a checkout at `rust2` and `rust3`, and `r2-stage1`/`r3-stage` linked):

<details>

```sh
#!/bin/sh
unset CARGO_TARGET_DIR

cargo new repro-76720
cd repro-76720
cargo add serde_derive@1.0.162

build() {
    worktree=$1
    commit=$2
    serde_version=$3
    toolchain=$4

    # We can't use RTIM because this only repros with `omit-git-hash`.
    (cd ../$worktree && git checkout $commit && x build std proc_macro)
    cargo update -p serde_derive --precise $serde_version
    cargo clean -p repro-76720 -p serde_derive
    cargo +$toolchain-stage1 build
}

build rust2 6874f4e3fc2a16be7c78e702d068bbc1daa90e16 1.0.162 r2
build rust3 999ac5f7770bff68bd65f490990d32c3ec1faaa6 1.0.163 r3
```

</details>

going to work tomorrow on actually fixing the bug now that there's an MCVE :) ty again!

jyn514 committed:
Load only the header for crate_matches

Previously, we used the following info to determine whether to load the crate:
1. The METADATA_HEADER, which includes a METADATA_VERSION constant
2. The embedded rustc version
3. Various metadata in the `CrateRoot`, including the SVH

This worked ok most of the time. Unfortunately, when building locally the rustc version is always
the same because `omit-git-hash` is on by default. That meant that we depended only on 1 and 3, and
we are not very good about bumping METADATA_VERSION (it's currently at 7) so in practice we were
only depending on 3. `CrateRoot` is a very large struct and changes somewhat regularly, so this led
to a steady stream of crashes from trying to load it.

Change the logic to add an intermediate step between 2 and 3: introduce a new `CrateHeader` struct
that contains only the minimum info needed to decide whether the crate should be loaded or not. That
avoids having to load all of `CrateRoot`, which in practice means we should crash much less often.

Note that this works because the SVH should be different between any two dependencies, even if the
compiler has changed, because we use `-Zbinary-dep-depinfo` in bootstrap. See
https://github.com/rust-lang/rust/pull/111329#issuecomment-1538303474 for more details about how the
original crash happened.

jyn514 commented:
> going to work tomorrow on actually fixing the bug now that there's an MCVE :)

Ok, I tried this just now and it doesn't actually work - it *does* successfully load the header without crashing, but because all of the (name, triple, svh hash, is_proc_macro) data is the same, it actually *doesn't* reject the crate and continues to try and load it.

I think we need to make sure that changes to the layout of `CrateRoot` are embedded into the SVH hash. Looks like it's determined here today: https://github.com/rust-lang/rust/blob/3a75f10af12e959273d19cf0f867f066e3db8add/compiler/rustc_middle/src/hir/map/mod.rs#L1147-L1174

I am not quite sure where to go from here - hashing the entire contents of `CrateRoot` seems expensive and unnecessary, but the only other thing I can think of to do is to hash `mem::size_of::<CrateRoot>`. @bjorn3 @Mark-Simulacrum does that seem like an ok change? I think it will work in the majority of cases, but it's not a reliable fix.

jyn514 commented:
ah wait no I must be doing something wrong - the version of serde_derive has changed, so the hash *must* be different. Let me figure out what I broke.

jyn514 commented:
mmm wait no I was right the first time - even though `serde_derive` has changed, the version of `quote` (which it depends on) has not.

jyn514 commented:
> hashing the entire contents of CrateRoot seems expensive and unnecessary, but the only other thing I can think of to do is to hash mem::size_of::<CrateRoot>

really what we want is something like StableTypeId, but i think TypeId intentionally varies across invocations :/

jyn514 commented:
> > hashing the entire contents of CrateRoot seems expensive and unnecessary, but the only other thing I can think of to do is to hash mem::size_of::
> 
> really what we want is something like StableTypeId, but i think TypeId intentionally varies across invocations :/

this is a terrible idea, but RustcEncodable could also implement that trait for us by hashing std::any::type_name and size_of recursively 🙃

Mark-Simulacrum commented:
I'm a little confused by the current state of your code / what is still broken. Are you saying that we correctly reject the older copy of serde_derive, but when loading quote (which has the same version) we are failing to reject the older version of that crate? My expectation is that Cargo/rustc should be able to uniquely identify a dependency by something (e.g., SVH) -- there shouldn't be a case where we're loading the wrong version if we've deserialized the header correctly. Is that not the case?

Mark-Simulacrum commented:
To elaborate a little more: if the two files aren't distinguished by something in that header, I would expect that the 2nd compilation would overwrite the previous file, so we should still be fine. This is how things normally work out after recompiles, I think, in your dependencies: we never start a rustc until the deps it's going to try and load are refreshed (either into the same file or a sufficiently distinguishable file through the metadata header).

jyn514 commented:
> expectation is that Cargo/rustc should be able to uniquely identify a dependency by something (e.g., SVH) -- there shouldn't be a case where we're loading the wrong version if we've deserialized the header correctly.

We are able to uniquely *identify* the dependency, but the metadata format has changed between the two versions of the compiler used to build it, so when we go to load it, we crash because we can't deserialize CrateRoot.

> if the two files aren't distinguished by something in that header, I would expect that the 2nd compilation would overwrite the previous file, so we should still be fine. 

Cargo does not know that the compiler changed, so it isn't recompiling `quote` a second time. Are you saying -Zbinary-dep-depinfo should let it know to recompile? I wasn't able to replicate that in the MCVE since I was using different worktrees for the commits - not at a computer now, but I can try using the same worktree for both and see if it helps.

Mark-Simulacrum commented:
Why are we loading the wrong file? Is there a *right* file at that point in time (like for serde derive in my writeup)?

Cargo should be distinguishing and recompiling everything - the cause of the error I investigated was that even though we knew the right crate hash, we didn't get to actually knowing the crate hash of the file we were trying to load due to differences in the metadata format. My proposal should have fixed that though.

jyn514 commented:
> Cargo does not know that the compiler changed, so it isn't recompiling `quote` a second time. Are you saying -Zbinary-dep-depinfo should let it know to recompile? I wasn't able to replicate that in the MCVE since I was using different worktrees for the commits - not at a computer now, but I can try using the same worktree for both and see if it helps.

With `-Zbinary-dep-depinfo` and a shared worktree, I can't reproduce the crash any more outside rust-analyzer-proc-macro-srv, even without any changes at all to the compiler :/

<details>

```
#!/bin/sh
unset CARGO_TARGET_DIR CARGO_REGISTRIES_CRATES_IO_PROTOCOL

cd $(dirname $0)
cargo new repro-76720
cd repro-76720
cargo add serde_derive@1.0.162
cargo clean

build_rustc() {
    worktree=$1
    commit=$2
    # We can't use RTIM because this only repros with `omit-git-hash`.
    (cd ../$worktree && git checkout $commit && x build std proc_macro)
}

build() {
    worktree=$1
    commit=$2
    serde_version=$3
    toolchain=$4

    build_rustc $worktree $commit
    cargo update -p serde_derive --precise $serde_version
    cargo +$toolchain-stage1 build -Zbinary-dep-depinfo
}

build rust2 71f6675de1faa2a8e897a8faaa861a0157b0e355 1.0.162 r2
build rust2 51cb5614dde163c53aab73e2dc33bb9f388b50a8 1.0.163 r2
```

</details>

jyn514 commented:
Ok, I stopped using the extern minimization, which was giving misleading results. The new script I'm using to test is

```sh
#!/bin/sh
set -e

unset CARGO_TARGET_DIR CARGO_REGISTRIES_CRATES_IO_PROTOCOL

git checkout 6874f4e3fc2a16be7c78e702d068bbc1daa90e16
git cherry-pick metadata-ice
rm -rf build/host/stage*
x build # --stage 1 proc-macro-srv-cli
git checkout 999ac5
git cherry-pick metadata-header
x build # --stage 1 proc-macro-srv-cli
```

which is exactly the MCVE from https://github.com/rust-lang/rust/issues/76720#issuecomment-1537546470 plus some gunk at the start from env variables I have set locally. If you comment out the `cherry-pick` commands it ICEs when building proc-macro-srv-cli; if you add them back, it passes successfully.

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/5087314773/jobs/9142570849) [(plain)](https://github.com/rust-lang/rust/commit/98e41bd90f2d8a0791936935c879301eccf59d2b/checks/13776719989/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Prepare all required actions
Getting action download info
Download action repository 'actions/checkout@v3' (SHA:8e5e7e5ab8b370d6c329ec480221332ada57f0ab)
Download action repository 'rust-lang/simpleinfra@master' (SHA:cdb59432ef9e018dd6db5879e720eccc3289da68)
Complete job name: PR - mingw-check-tidy
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
env:
  CI_JOB_NAME: mingw-check-tidy
---
Building wheels for collected packages: reuse
  Building wheel for reuse (pyproject.toml): started
  Building wheel for reuse (pyproject.toml): finished with status 'done'
  Created wheel for reuse: filename=reuse-1.1.0-cp310-cp310-manylinux_2_35_x86_64.whl size=180115 sha256=c3e1cf75e5aea7348ad201d16633573bc9f3aa58c8e29a6ca459f306300d6c9d
  Stored in directory: /tmp/pip-ephem-wheel-cache-kgewhqxh/wheels/c2/3c/b9/1120c2ab4bd82694f7e6f0537dc5b9a085c13e2c69a8d0c76d
Installing collected packages: boolean-py, binaryornot, setuptools, reuse, python-debian, markupsafe, license-expression, jinja2, chardet
  Attempting uninstall: setuptools
    Found existing installation: setuptools 59.6.0
    Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
---
Successfully tagged rust-ci:latest
Built container sha256:38781af1e8afdc2b6ee4c758023f29e682b370285bc446474cae4810a4469b16
Uploading finished image to https://ci-caches.rust-lang.org/docker/e51156f19850ce886cec818e46dc2a021e0aa7c270d15673e8fe74cd8522fc8ac3995109aebb688ee49ed586735ed4cf5f8c06d44c48298fa09c35ae4b082281

<botocore.awsrequest.AWSRequest object at 0x7fb035eeff50>
gzip: stdout: Broken pipe
xargs: docker: terminated by signal 13
[CI_JOB_NAME=mingw-check-tidy]
[CI_JOB_NAME=mingw-check-tidy]
---
   Compiling tidy v0.1.0 (/checkout/src/tools/tidy)
    Finished release [optimized] target(s) in 17.18s
##[endgroup]
fmt check
Diff in /checkout/compiler/rustc_metadata/src/locator.rs at line 670:
         if header.is_proc_macro_crate != self.is_proc_macro {
             info!(
                 "Rejecting via proc macro: expected {} got {}",
-                self.is_proc_macro,
-                header.is_proc_macro_crate,
+                self.is_proc_macro, header.is_proc_macro_crate,
             return None;
         }
         }
Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/compiler/rustc_metadata/src/rmeta/table.rs" "/checkout/compiler/rustc_metadata/src/rmeta/decoder.rs" "/checkout/compiler/rustc_metadata/src/dependency_format.rs" "/checkout/compiler/rustc_metadata/src/native_libs.rs" "/checkout/compiler/rustc_metadata/src/lib.rs" "/checkout/compiler/rustc_metadata/src/locator.rs" "/checkout/compiler/rustc_metadata/src/creader.rs" "/checkout/compiler/rustc_metadata/src/rmeta/def_path_hash_map.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
```

</details>

bjorn3 commented on compiler/rustc_metadata/src/rmeta/decoder.rs:
```diff
@@ -73,7 +73,7 @@ pub(crate) struct CrateMetadata {
     blob: MetadataBlob,
 
-    // --- Some data pre-decoded from the metadata blob, usually for performance ---
+    /// --- Some data pre-decoded from the metadata blob, usually for performance ---
```
I think this was meant as section header, not as documentation for the `root` field.

bjorn3 commented on compiler/rustc_metadata/src/rmeta/mod.rs:
```diff
@@ -199,7 +199,27 @@ pub(crate) struct ProcMacroData {
+/// Prefer [`CrateRoot`] whenever possible to avoid ICEs when using `omit-git-hash` locally.
+/// See #76720 for more details.
+///
+/// If you do modify this struct, also bump the [`METADATA_VERSION`] constant.
```
Should there be a `METADATA_VERSION` bump in this PR?

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-14`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/5090945772/jobs/9150377659) [(plain)](https://github.com/rust-lang/rust/commit/62314be705aee343870f3bc36cb5eb7e89ef6269/checks/13786590716/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
error: test failed, to rerun pass `--lib`

failures:

---- abis::test_version_check stdout ----
thread 'abis::test_version_check' panicked at 'called `Result::unwrap()` on an `Err` value: Custom { kind: InvalidData, error: "unsupported metadata version 8" }', crates/proc-macro-srv/src/abis/mod.rs:144:55

---- tests::list_test_macros stdout ----
---- tests::list_test_macros stdout ----
thread 'tests::list_test_macros' panicked at 'called `Result::unwrap()` on an `Err` value: "Cannot create expander for /checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/build/proc-macro-test-b10b33ad22ec78c7/out/target/debug/libproc_macro_test_impl.so: unsupported metadata version 8"', crates/proc-macro-srv/src/tests/utils.rs:39:44
---- tests::test_derive_empty stdout ----
---- tests::test_derive_empty stdout ----
thread 'tests::test_derive_empty' panicked at 'called `Result::unwrap()` on an `Err` value: Io(Custom { kind: InvalidData, error: "unsupported metadata version 8" })', crates/proc-macro-srv/src/tests/utils.rs:28:48
---- tests::test_attr_macro stdout ----
---- tests::test_attr_macro stdout ----
thread 'tests::test_attr_macro' panicked at 'called `Result::unwrap()` on an `Err` value: Io(Custom { kind: InvalidData, error: "unsupported metadata version 8" })', crates/proc-macro-srv/src/tests/utils.rs:28:48
---- tests::test_derive_error stdout ----
---- tests::test_derive_error stdout ----
thread 'tests::test_derive_error' panicked at 'called `Result::unwrap()` on an `Err` value: Io(Custom { kind: InvalidData, error: "unsupported metadata version 8" })', crates/proc-macro-srv/src/tests/utils.rs:28:48
---- tests::test_fn_like_macro_clone_ident_subtree stdout ----
---- tests::test_fn_like_macro_clone_ident_subtree stdout ----
thread 'tests::test_fn_like_macro_clone_ident_subtree' panicked at 'called `Result::unwrap()` on an `Err` value: Io(Custom { kind: InvalidData, error: "unsupported metadata version 8" })', crates/proc-macro-srv/src/tests/utils.rs:28:48
---- tests::test_fn_like_macro_noop stdout ----
---- tests::test_fn_like_macro_noop stdout ----
thread 'tests::test_fn_like_macro_noop' panicked at 'called `Result::unwrap()` on an `Err` value: Io(Custom { kind: InvalidData, error: "unsupported metadata version 8" })', crates/proc-macro-srv/src/tests/utils.rs:28:48

---- tests::test_fn_like_macro_clone_literals stdout ----
---- tests::test_fn_like_macro_clone_literals stdout ----
thread 'tests::test_fn_like_macro_clone_literals' panicked at 'called `Result::unwrap()` on an `Err` value: Io(Custom { kind: InvalidData, error: "unsupported metadata version 8" })', crates/proc-macro-srv/src/tests/utils.rs:28:48
---- tests::test_fn_like_macro_clone_raw_ident stdout ----
---- tests::test_fn_like_macro_clone_raw_ident stdout ----
thread 'tests::test_fn_like_macro_clone_raw_ident' panicked at 'called `Result::unwrap()` on an `Err` value: Io(Custom { kind: InvalidData, error: "unsupported metadata version 8" })', crates/proc-macro-srv/src/tests/utils.rs:28:48
---- tests::test_fn_like_mk_idents stdout ----
---- tests::test_fn_like_mk_idents stdout ----
thread 'tests::test_fn_like_mk_idents' panicked at 'called `Result::unwrap()` on an `Err` value: Io(Custom { kind: InvalidData, error: "unsupported metadata version 8" })', crates/proc-macro-srv/src/tests/utils.rs:28:48
---- tests::test_fn_like_mk_literals stdout ----
---- tests::test_fn_like_mk_literals stdout ----
thread 'tests::test_fn_like_mk_literals' panicked at 'called `Result::unwrap()` on an `Err` value: Io(Custom { kind: InvalidData, error: "unsupported metadata version 8" })', crates/proc-macro-srv/src/tests/utils.rs:28:48

failures:
    abis::test_version_check
    tests::list_test_macros
```

</details>

bjorn3 commented:
https://github.com/rust-lang/rust/blob/917b0b6c70f078cb08bbb0080c9379e4487353c3/src/tools/rust-analyzer/crates/proc-macro-api/src/version.rs#L125 will need to be changed to accept both 7 and 8 as metadata version and the change will need to be synced back to rust-analyzer. Another option would be to put a version as first field of the `CrateHeader` to change whenever the format of `CrateHeader` changes. The `METADATA_VERSION` can then be kept changing only when the encoding of the rustc version string changes. If you were to start the version in the `CrateHeader` with say `0xffff_u16` it would not conflict with any recent past versions of the `CrateRoot` encoding as there is no target shipped with rustc where the length of the name has an ULEB128 value starting with something as big as `0xffff`.

jyn514 commented:
I'm going to reuse METADATA_VERSION instead of adding a second version constant for now. I think it will be very unlikely in practice to need to change `CrateHeader`.

jyn514 committed:
Update proc-macro-api for the new rustc metadata format

bjorn3 commented:
Do note that this will break rust-analyzer proc macro loading on nightly until this change is synced back and I believe there was some issue preventing syncs back right now.

Veykril commented:
No that should be fine. Those checks are only done at the proc-macro server side now, the rust-analyzer server doesn't check the dylibs itself anymore.

jyn514 commented:
@Veykril who handles subtree syncs? Will you or another RA member make a sync after this lands or should I do the sync myself?

Veykril commented:
@lnicola usually does them, but we currently have a problem with syncing because it would break our CI atm. I forgot to look into that, I'll look into that again in a bit.

bjorn3 commented:
`MacroDylib::new` calls `version::read_dylib_info`. `MacroDylib::new` is called at https://github.com/rust-lang/rust/blob/917b0b6c70f078cb08bbb0080c9379e4487353c3/src/tools/rust-analyzer/crates/rust-analyzer/src/reload.rs#L651 which is part of the rust-analyzer binary, not the rust-analyzer-proc-macro-srv binary. Or am I missing something?

jyn514 commented:
I wonder if RA should get the rustc version from proc-macro-srv instead of trying to read it directly. Then it wouldn't be broken by these updates. The only time it needs the version is if it failed to load the proc-macro anyway, which should be very rare now that it's using the server in the sysroot.

Veykril commented:
> `MacroDylib::new` calls `version::read_dylib_info`. `MacroDylib::new` is called at
> 
> https://github.com/rust-lang/rust/blob/917b0b6c70f078cb08bbb0080c9379e4487353c3/src/tools/rust-analyzer/crates/rust-analyzer/src/reload.rs#L651
> which is part of the rust-analyzer binary, not the rust-analyzer-proc-macro-srv binary. Or am I missing something?

Well, it doesn't do that anymore on rust-lang/rust-analyzer, the subtree is just that outdated 😅

bjorn3 commented:
@bors r+

bors commented:
:pushpin: Commit 60e95e76d059e3bbeb19d819914df3bcf7684a82 has been approved by `bjorn3`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=bjorn3 60e95e76d059e3bbeb19d819914df3bcf7684a82 -->
<!-- homu: {"type":"Approved","sha":"60e95e76d059e3bbeb19d819914df3bcf7684a82","approver":"bjorn3","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 60e95e76d059e3bbeb19d819914df3bcf7684a82 with merge 99ff5afeb8717c096a4f153896e0c3856a3934ad...
<!-- homu: {"type":"BuildStarted","head_sha":"60e95e76d059e3bbeb19d819914df3bcf7684a82","merge_sha":"99ff5afeb8717c096a4f153896e0c3856a3934ad"} -->
